framework/math/ParallelVector/ParallelVector.h

```diff
@@ -0,0 +1,164 @@
+   * the given communicator whose entries are set to zero.
+   */
+  ParallelVector(uint64_t local_size,
+                 uint64_t global_size,
+                 MPI_Comm communicator);
+
+  /**Copy constructor.*/
+  ParallelVector(const ParallelVector& other);
+
+  /**Move constructor.*/
+  ParallelVector(ParallelVector&& other) noexcept;
+
+  virtual std::unique_ptr<ParallelVector> MakeCopy() const = 0;
+  virtual std::unique_ptr<ParallelVector> MakeNewVector() const = 0;
+
+  /** Returns the raw stl-vector associated with the local data of this
+   * vector.*/
+  virtual std::vector<double>& RawValues() = 0;
+
+  /// Return the size of the locally owned portion of the parallel vector.
+  uint64_t LocalSize() const { return local_size_; }
+
+  /// Return the global size of the parallel vector.
+  uint64_t GlobalSize() const { return global_size_; }
+
+  /**
+   * Read only accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double operator[](int64_t local_id) const = 0;
+
+  /**
+   * Read/write accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double& operator[](int64_t local_id) = 0;
+
+  /// Return a vector containing the locally owned data.
+  virtual std::vector<double> MakeLocalVector() = 0;
+
+  /**
+   * Set the entries of the locally owned portion of the parallel vector
+   * to the given value.
+   */
+  virtual void Set(double value) = 0;
+
+  /**
+   * Set the entries of the locally owned portion of the parallel vector
+   * to the given STL vector.
+   */
+  virtual void Set(const std::vector<double>& local_vector) = 0;
+
+  /**
+   * Define a set or add operation for the given global id-value pair
+   *
+   * This routine adds the global id-value pair to the set operation cache,
+   * which upon execution of Assemble, communicates the operations to the
+   * appropriate process.
+   */
+  virtual void SetValue(int64_t global_id, double value, VecOpType op_type) = 0;
+
+  /**
+   * Group multiple operations into a single call.
+   *
+   * This routine goes through the given global id-value pairs and calls
+   * SetValue for each.
+   */
+  virtual void SetValues(const std::vector<int64_t>& global_ids,
+                         const std::vector<double>& values,
+                         VecOpType op_type) = 0;
+
+  /**In place adding of vectors. The sizes must be compatible.*/
+  virtual void operator+=(const ParallelVector& y) = 0;
+
+  /**Adds a vector multiplied by scalar a. Optimized for a=1.0 and -1.0*/
+  virtual void PlusAY(double a, const ParallelVector& y) = 0;
```

Comment:
```suggestion
  virtual void Add(const ParallelVector& y, const double a = 1.0) = 0;
```
What about this with a check for `a = 1.0` to avoid unnecessary multiplication? `PlusAY` doesn't scream add a scaled vector but rather add a matrix vector product. I think avoiding using math notation in function names is better than the alternative. With case rules, it makes it even less clear.

Reply:
Ok. We can do this. I want to eventually get to all of these:
<img width="705" alt="image" src="https://github.com/chi-tech/chi-tech/assets/25042529/a82c04f7-6ad2-4c3f-a04d-c5ec617564ad">

Reply:
The overall theme here, and the comments below, seem to be to favor overloads. We should discuss this but my opinion is that it is much harder to search through the documentation of overloads than it is to look for hints of the intent through the method name. You basically get additional short-hand information folded into the name of the method.

In this case I don't think `Add` here is a necessarily a bad option. I just think that adding more functionality where the statement reads like an equation is better. Examples:
```
x.Add(y, 2.0) --> x add y 2.0
x.PlusAy(2.0, y) --> x plus ay, a=2.0

x.Add({y1,y2,y3}, {1.0,2.0,3.0})
x.PlusMultiAY({1.0,2.0,3.0}, {y1,y2,y3})
```
as long as we are consistent the A one can eventually get a feel for the method name without looking it up (Like for PETSc).

Reply:
<img width="697" alt="image" src="https://github.com/chi-tech/chi-tech/assets/25042529/7e6b03f5-0ae9-448f-b336-e652f529f6b4">

Reply:
Okay, so I am sensing that we want to reimplement most/all the PETSc functions with effectively the same names. If that is the case, put the actual equations being carried out in the documentation as PETSc does. This helps a user who isn't used to `AY` being a scalar times a vector. Also, if someone is trying to add vectors, I would imaging a user looks up `Add` over `Plus`, but that's just me. I'm fine with whatever convention you choose here, but I would add equations to documentation.